pandas/io/excel/_base.py

```diff
@@ -346,33 +353,33 @@
 @Appender(_read_excel_doc)
 def read_excel(
     io,
-    sheet_name=0,
-    header=0,
+    sheet_name: str | int | list[int] | list[str] | None = 0,
+    header: int | Sequence[int] | None = 0,
     names=None,
-    index_col=None,
+    index_col: int | Sequence[int] | None = None,
     usecols=None,
-    squeeze=None,
+    squeeze: bool | None = None,
     dtype: DtypeArg | None = None,
-    engine=None,
+    engine: Literal["xlrd", "openpyxl", "odf", "pyxlsb"] | None = None,
     converters=None,
-    true_values=None,
-    false_values=None,
-    skiprows=None,
-    nrows=None,
+    true_values: Iterable[Hashable] | None = None,
+    false_values: Iterable[Hashable] | None = None,
+    skiprows: Sequence[int] | int | Callable[[int], object] | None = None,
+    nrows: int | None = None,
     na_values=None,
-    keep_default_na=True,
-    na_filter=True,
-    verbose=False,
+    keep_default_na: bool = True,
+    na_filter: bool = True,
+    verbose: bool = False,
     parse_dates=False,
     date_parser=None,
-    thousands=None,
-    decimal=".",
-    comment=None,
-    skipfooter=0,
-    convert_float=None,
-    mangle_dupe_cols=True,
+    thousands: str | None = None,
+    decimal: str = ".",
+    comment: str | None = None,
+    skipfooter: int = 0,
+    convert_float: bool | None = None,
+    mangle_dupe_cols: bool = True,
     storage_options: StorageOptions = None,
-):
+) -> DataFrame | dict[str, DataFrame] | dict[int, DataFrame]:
```

Comment:
It would be great to have an overload (but a bit lengthy):

```py
IntStrT = TypeVar("IntStrT", int, str) # or have a third overload

@overload
def read_excel(..., sheet_name: str | int = ..., ...) -> DataFrame: ...
@overload
def read_excel(..., sheet_name: list[IntStrT] = ..., ...) -> dict[IntStrT, DataFrame]: ...

def read_excel(..., sheet_name: str | int | list[int] | list[str]= ..., ...) -> DataFrame | dict[str, DataFrame] | dict[int, DataFrame]: ...
```

Reply:
Hm you are right. Wanted to do that after we disallow positional arguments but is not necessary here

Reply:
Will add another overload, because list[int | str] is not allowed, they have to be either ints or strings

Reply:
I might be wrong: the TypeVar will bind to either int or str not both at the same time

Reply:
Ah, will check that

Reply:
Yep you are correct, thx. Pycharm does not recognize that, so thought this was the same as list[int | str]

Reply:
Using the typevar caused pyright to error with
```
An internal error occurred while type checking file "/home/runner/work/pandas/pandas/pandas/io/excel/_base.py": Error: Debug Failure. Unexpected flow node flags
    at narrowConstrainedTypeVarRecursive (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1218:17)
    at Object.narrowConstrainedTypeVar (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1223:41)
    at Object.narrowConstrainedTypeVar (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:22189:31)
    at x.visitReturn (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:806:74)
    at x.visitNode (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:390:26)
    at x.walk (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:97:37)
    at x.walk (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:244:19)
    at forEach (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:106:22)
    at Array.forEach (<anonymous>)
    at x.walkMultiple (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:104:15)
Error performing analysis: Error: Debug Failure. Unexpected flow node flags
    at narrowConstrainedTypeVarRecursive (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1218:17)
    at Object.narrowConstrainedTypeVar (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/codeFlowEngine.ts:1223:41)
    at Object.narrowConstrainedTypeVar (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:22189:31)
    at x.visitReturn (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:806:74)
    at x.visitNode (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:390:26)
    at x.walk (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:97:37)
    at x.walk (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/checker.ts:244:19)
    at forEach (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:106:22)
    at Array.forEach (<anonymous>)
    at x.walkMultiple (/opt/hostedtoolcache/node/16.13.1/x64/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/parseTreeWalker.ts:104:15)
```

remove it for now. Do you have any idea where this is from?

Reply:
Then go with a third overload for now. Can you create an issue at pyright?

Reply:
Yep added the overload already. Will try to create a minimal example to post there

Reply:
Opened https://github.com/microsoft/pyright/issues/2775

Reply:
Thanks for pushing that